Sources/Basics/Terminator.swift

```diff
@@ -0,0 +1,112 @@
+    private let registry = ThreadSafeKeyValueStore<String, (name: String, handler: TerminationHandler)>()
+    private let terminationQueue = DispatchQueue(label: "org.swift.swiftpm.terminator")
+
+    private let terminated = ThreadSafeBox<Bool>(false)
+
+    public init(
+        killTimeout: DispatchTimeInterval = .seconds(30),
+        observabilityScope: ObservabilityScope?
+    ) {
+        self.killTimeout = killTimeout
+        self.observabilityScope = observabilityScope
+    }
+
+    @discardableResult
+    public func register(name: String, handler: @escaping TerminationHandler) -> RegistrationKey {
+        if self.terminated.get() ?? false {
+            self.observabilityScope?.emit(debug: "not registering '\(name)' with terminator as its already cancelled")
+            // TODO: do something better in this case
+            return "noop" // nothing was done
+        }
+        let key = UUID().uuidString
+        self.observabilityScope?.emit(debug: "registering '\(name)' with terminator")
+        self.registry[key] = (name: name, handler: handler)
+        return key
+    }
+
+    public func register(_ process: TSCBasic.Process) -> RegistrationKey {
+        self.register(name: "\(process.arguments.joined(separator: " "))", handler: {
+            process.terminate(timeout: .now() + self.killTimeout)
+        })
+    }
+
+    public func deregister(_ key: RegistrationKey) {
+        self.registry[key] = nil
+    }
+
+    public func terminate() {
+        self.terminated.put(true)
+
+        self.observabilityScope?.emit(info: "starting termination cycle with \(self.registry.count) termination handlers registered")
+
+        let success = ThreadSafeArrayStore<String>()
+        let group = DispatchGroup()
+        for (_, (name, handler)) in self.registry.get() {
+            group.enter()
+            self.terminationQueue.async {
+                defer { group.leave() }
+                do {
+                    self.observabilityScope?.emit(debug: "terminating '\(name)'")
+                    try handler()
+                    success.append(name)
+                } catch {
+                    self.observabilityScope?.emit(warning: "failed terminating '\(name)': \(error)")
+                }
+            }
+        }
+
+        if case .timedOut = group.wait(timeout: .now() + self.killTimeout) {
+            self.observabilityScope?.emit(warning: "timeout waiting for termination with \(self.registry.count - success.count) termination handlers remaining")
+        } else {
+            self.observabilityScope?.emit(info: "termination cycle completed successfully")
+        }
+    }
+}
+
+extension TSCBasic.Process {
+    fileprivate func terminate(timeout: DispatchTime) {
+        // send graceful shutdown signal
+        self.signal(SIGINT)
+
+        // start a thread to see if we need to terminate more forcibly
+        let forceKillSemaphore = DispatchSemaphore(value: 1)
+        let forceKillThread = TSCBasic.Thread {
```

Comment:
Is the reason for using a thread and not a dispatch async that we can't know if a dispatch loop is running?

Reply:
since this is likely canceling threads that have been scheduled with Dispatch, it gives us a bit of a separation between the systems. dispatch does work to minimize how much threads it spawns so there is some risk the shutdown thread will not be spawned or prioritized if we use out global dispatch queue

Reply:
I see — good points, thanks for explaining.